File: Examples/GroupDocs.Annotation.Examples.CSharp/GroupDocs.AnnotationDataStorage/RepositoryPathFinder.cs
using System;
using System.Configuration;
using System.IO;
using System.Web;

namespace GroupDocs.Data.Json
{
    public interface IRepositoryPathFinder
    {
        string Find(string repository);
    }

    public class RepositoryPathFinder : IRepositoryPathFinder
    {
        public const string RepoBasePathKey = "GroupDocsRepoBasePath";

        public string Find(string repository)
        {
            var ctx = HttpContext.Current;
            if (ctx == null)
            {
                return GetConfiguredPath(repository);
            }

            var basePath = (ctx.Application != null ? (string) ctx.Application[RepoBasePathKey] : null);
            return (string.IsNullOrWhiteSpace(basePath) ?
                (ctx.Server != null ?
                    ctx.Server.MapPath(string.Format("~/App_Data/{0}", repository)) : GetConfiguredPath(repository)) :
                Path.Combine(basePath, repository));
        }

        private static string GetConfiguredPath(string repository)
        {
            return Path.Combine(
                ConfigurationManager.AppSettings[RepoBasePathKey] ?? AppDomain.CurrentDomain.SetupInformation.ApplicationBase,
                repository);
        }
    }
}

File: Examples/GroupDocs.Annotation.Examples.VisualBasic/GroupDocs.AnnotationDataStorage/RepositoryPathFinder.vb
Imports System.IO
Imports System.Web
Imports System.Configuration

Public Interface IRepositoryPathFinder
    Function Find(repository As String) As String
End Interface

Public Class RepositoryPathFinder
    Implements IRepositoryPathFinder
    Public Const RepoBasePathKey As String = "GroupDocsRepoBasePath"

    Public Function Find(repository As String) As String Implements IRepositoryPathFinder.Find
        Dim ctx = HttpContext.Current
        If ctx Is Nothing Then
            Return GetConfiguredPath(repository)
        End If

        Dim basePath = (If(ctx.Application IsNot Nothing, DirectCast(ctx.Application(RepoBasePathKey), String), Nothing))
        Return (If(String.IsNullOrWhiteSpace(basePath), (If(ctx.Server IsNot Nothing, ctx.Server.MapPath(String.Format("~/App_Data/{0}", repository)), GetConfiguredPath(repository))), Path.Combine(basePath, repository)))
    End Function

    Private Shared Function GetConfiguredPath(repository As String) As String
        Return Path.Combine(If(ConfigurationManager.AppSettings(RepoBasePathKey), AppDomain.CurrentDomain.SetupInformation.ApplicationBase), repository)
    End Function
End Class
